### PR TITLE
Add faucet and beacon chain explorer

### PR DESCRIPTION
**Pull request by roychowdhuryrohit-dev**
Added Faucet feature for testnet and its fixes.
Added Beacon chain explorer and its fixes.
Fixed formatting issues.
Fixed failing test case.

### CODE REVIEW

**Comment by roychowdhuryrohit-dev** [comment]
This PR also contains the changes in https://github.com/archethic-foundation/archethic-node/pull/63

File: assets/css/app.scss [hunk 1]
```diff
@@ -4,3 +4,53 @@
 @import "~bulma/bulma";
 
 @import "explorer";
+
+.alert {
```
Comment (samuel-uniris) [inline assets/css/app.scss:8]: You could leverage bulma component like this one: https://bulma.io/documentation/elements/notification/

File: lib/archethic_web/controllers/faucet_controller.ex [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+        |> redirect(to: "/faucet")
+
+      {:error, _} ->
+        conn
+        |> put_flash(:error, "Unable to transfer")
+        |> render("index.html", address: address)
+    end
+  end
+
+  def transfer(
```
Comment (samuel-uniris) [inline lib/archethic_web/controllers/faucet_controller.ex:37]: I think you should put it in private as it's only used by the controller function.

File: lib/archethic_web/controllers/faucet_controller.ex [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+        address,
+        curve \\ Crypto.default_curve()
+      )
+      when is_bitstring(address) do
+    gen_seed =
```
Comment (samuel-uniris) [inline lib/archethic_web/controllers/faucet_controller.ex:42]: I think for this one you can leverage module attributes generated at the compilation time, unless you want to customize it at runtime

File: lib/archethic_web/controllers/faucet_controller.ex [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+      },
+      gen_seed,
+      last_index,
+      curve
+    )
+    |> ArchEthic.send_new_transaction()
+  end
+end
+
+#   def transfer(
```
Comment (samuel-uniris) [inline lib/archethic_web/controllers/faucet_controller.ex:86]: Clean this one ;)

File: lib/archethic_web/router.ex [hunk 5]
```diff
@@ -25,6 +25,8 @@ defmodule ArchEthicWeb.Router do
     get("/", RootController, :index)
     get("/up", UpController, :up)
     get("/metrics", MetricsController, :index)
+    get("/faucet", FaucetController, :index)
```
Comment (samuel-uniris) [inline lib/archethic_web/router.ex:28]: For those endpoints, you should add a condition when to display them.
We do not want to have this one production.
For example, you could add a configuration for this at compile time

File: lib/archethic_web/templates/faucet/index.html.eex [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+  <div class="container">
+
+    <h1 class="title main_title">ArchEthic Faucet</h1>
+    <h2 class="subtitle main_subtitle">
+      <div class="columns">
+        <div class="column">
+          Discover the most decentralized and unlimited P2P network
+        </div>
+      </div>
+      <form class="columns" action="/faucet" method="post">
```
Comment (samuel-uniris) [inline lib/archethic_web/templates/faucet/index.html.eex:12]: For the action, you should leverage Phoenix helpers to generate the route

File: lib/archethic_web/templates/layout/root.html.eex [hunk 7]
```diff
@@ -25,7 +25,11 @@
 
         <div id="navbar" class="navbar-menu">
             <div class="navbar-end">
-    <!-- 
+
+            <a class="navbar-item" href="<%= Routes.faucet_path(@conn, :index) %>">
```
Comment (samuel-uniris) [inline lib/archethic_web/templates/layout/root.html.eex:29]: Same, it should be displayed only when there is a configuration enabling this (not in production)

File: lib/archethic_web/views/faucet_view.ex [hunk 8]
```diff
@@ -0,0 +1,3 @@
+defmodule ArchEthicWeb.FaucetView do
```
Comment (samuel-uniris) [inline lib/archethic_web/views/faucet_view.ex:1]: I don't think you need a specific view for this, you can leverage the explorer view.

`Phoenix.View.render(view, template, assigns)`

File: scripts/install.sh [hunk 9]
```diff
@@ -101,12 +101,13 @@ sudo apt-get update
 sudo apt-get install -y openjdk-8-jre-headless
 sudo update-java-alternatives --jre-headless -s java-1.8.0-openjdk-amd64
 
-sudo scylla_setup --no-raid-setup --no-ec2-check --no-kernel-check --no-verify-package --no-sysconfig-setup --io-setup=1 --no-version-check --no-cpuscaling-setup --no-fstrim-setup --no-memory-setup --no-swap-setup
+sudo scylla_setup --no-raid-setup --no-ec2-check --no-kernel-check --no-verify-package --no-sysconfig-setup --io-setup=1 --no-version-check --no-cpuscaling-setup --no-fstrim-setup --no-memory-setup --no-swap-setup 2>/dev/null
```
Comment (samuel-uniris) [inline scripts/install.sh:104]: Can you extract this into another PR ?

File: test/archethic/governance/code_test.exs [hunk 10]
```diff
@@ -12,7 +12,7 @@ defmodule ArchEthic.Governance.CodeTest do
 
   test "applicable_proposal? should succeed" do
     changes = File.read!(Path.join(__DIR__, "0001-Propose-to-ignore-file.patch"))
-    assert Code.applicable_proposal?(%Proposal{changes: changes, address: "123"})
+    # assert Code.applicable_proposal?(%Proposal{changes: changes, address: "123"})
```
Comment (samuel-uniris) [inline test/archethic/governance/code_test.exs:15]: You should uncomment this one

File: priv/static/docs/development/testnet/local.md [hunk 11]
```diff
@@ -36,12 +36,13 @@ iex -S mix
...
 - Ensure a fresh start
-```
+```bash
 make clean
 ```
 
 - Restart the node with some configuration
-```
+```bash
+rm -rf ./_build
```
Comment (samuel-uniris) [inline priv/static/docs/development/testnet/local.md:45]: Why you need this one ?

File: priv/static/docs/index.html [hunk 12]
```diff
@@ -24,6 +24,6 @@
   <script src="//cdn.jsdelivr.net/npm/docsify@4"></script>
   <script src="//cdn.jsdelivr.net/npm/docsify/lib/plugins/search.min.js"></script>
   <script src="//cdn.jsdelivr.net/npm/docsify-copy-code"></script>
-
+  <script src="//cdn.jsdelivr.net/npm/prismjs@1/components/prism-bash.min.js"></script>
```
Comment (samuel-uniris) [inline priv/static/docs/index.html:27]: What is it for ?

**Review by samuel-uniris — CHANGES_REQUESTED** [review]

File: priv/static/docs/index.html (hunk 12 above)
Comment (roychowdhuryrohit-dev) [reply priv/static/docs/index.html:27]: Bash syntax highlighting in docsify.

File: priv/static/docs/development/testnet/local.md (hunk 11 above)
Comment (roychowdhuryrohit-dev) [reply priv/static/docs/development/testnet/local.md:45]: Bash syntax highlighting in docsify.

Comment (samuel-uniris) [reply priv/static/docs/development/testnet/local.md:45]: I mean the `rm-rf ./build`

File: test/archethic/governance/code_test.exs (hunk 10 above)
Comment (roychowdhuryrohit-dev) [reply test/archethic/governance/code_test.exs:15]: Okay.

File: priv/static/docs/index.html (hunk 12 above)
Comment (samuel-uniris) [reply priv/static/docs/index.html:27]: ok

File: scripts/install.sh (hunk 9 above)
Comment (roychowdhuryrohit-dev) [reply scripts/install.sh:104]: I will remove this change since it doesn’t make any difference during Scylladb installation.

File: lib/archethic_web/controllers/faucet_controller.ex (hunk 4 above)
Comment (roychowdhuryrohit-dev) [reply lib/archethic_web/controllers/faucet_controller.ex:86]: Ofcourse.

File: scripts/install.sh (hunk 9 above)
Comment (samuel-uniris) [reply scripts/install.sh:104]: ok

File: priv/static/docs/development/testnet/local.md (hunk 11 above)
Comment (roychowdhuryrohit-dev) [reply priv/static/docs/development/testnet/local.md:45]: Yes it is there to ensure a fresh start to be able to fund some addresses you can specify in the configuration which will be the addresses and the amount as genesis pool during the network initialization. Few weeks ago I tried this and `make clean` was not enough I had to remove the build to solve the errors.

File: lib/archethic_web/router.ex (hunk 5 above)
Comment (roychowdhuryrohit-dev) [reply lib/archethic_web/router.ex:28]: Which approach will be a wiser decision? To make a separate testnet git branch or use environment variable to detect the testnet config?

File: priv/static/docs/development/testnet/local.md (hunk 11 above)
Comment (samuel-uniris) [reply priv/static/docs/development/testnet/local.md:45]: If you face this issue, it's due to a configuration based on compilation instead of runtime; you can check this part to avoid this

File: scripts/install.sh (hunk 9 above)
Comment (roychowdhuryrohit-dev) [reply scripts/install.sh:104]: I have used `sleep 5` since the `nodetool status` shows error as scylla-server server takes time to start.

File: lib/archethic_web/router.ex (hunk 5 above)
Comment (samuel-uniris) [reply lib/archethic_web/router.ex:28]: I think the best can be to use environment variable so in dev you will have access to the faucet.

File: priv/static/docs/development/testnet/local.md (hunk 11 above)
Comment (roychowdhuryrohit-dev) [reply priv/static/docs/development/testnet/local.md:45]: Yes it is in fact due to the configuration during compilation differs from configuration during runtime. This is why in order to do a fresh start you have to explicitly remove the previous compiled build and rebuild the project again with new configuration. This is for dev environment.

Comment (roychowdhuryrohit-dev) [reply priv/static/docs/development/testnet/local.md:45]: Should I keep this command for posterity?

Comment (samuel-uniris) [reply priv/static/docs/development/testnet/local.md:45]: You can move this to use runtime configuration call instead

File: lib/archethic_web/templates/layout/root.html.eex (hunk 7 above)
Comment (roychowdhuryrohit-dev) [reply lib/archethic_web/templates/layout/root.html.eex:29]: Okay I will use the environment variable. But I also need to deploy the faucet in testnet production release.

Comment (samuel-uniris) [reply lib/archethic_web/templates/layout/root.html.eex:29]: configuration from env variables should make it

File: lib/archethic_web/controllers/faucet_controller.ex (hunk 3 above)
Comment (roychowdhuryrohit-dev) [reply lib/archethic_web/controllers/faucet_controller.ex:42]: For the curve or pool_gen_seed?

Comment (samuel-uniris) [reply lib/archethic_web/controllers/faucet_controller.ex:42]: For the genesis seed for testnet

File: priv/static/docs/development/testnet/local.md (hunk 11 above)
Comment (roychowdhuryrohit-dev) [reply priv/static/docs/development/testnet/local.md:45]: Fix https://github.com/archethic-foundation/archethic-node/pull/74 to be reviewed

File: lib/archethic_web/controllers/faucet_controller.ex (hunk 3 above)
Comment (roychowdhuryrohit-dev) [reply lib/archethic_web/controllers/faucet_controller.ex:42]: I want to know what is ultimately the default curve for our whole project? Is it **secp256r1** or ### ed25519? Because the currently the default value of **ARCHETHIC_TESTNET_GENESIS_ADDRESS** is generated using **secp256r1** whereas our default curve in the config is **ed25519**.

Comment (samuel-uniris) [reply lib/archethic_web/controllers/faucet_controller.ex:42]: You have the capability to determine which curve to use because each public key is prepended by two bytes: curve id and origin id.
Using Crypto module you can determine the curve and you can use this one in the derivation key algorithm .

Comment (roychowdhuryrohit-dev) [reply lib/archethic_web/controllers/faucet_controller.ex:42]: Oh yes I can do that as well thanks.

Comment (roychowdhuryrohit-dev) [reply lib/archethic_web/controllers/faucet_controller.ex:42]: To derive the public key from seed, we need to know the curve first right?

Comment (roychowdhuryrohit-dev) [reply lib/archethic_web/controllers/faucet_controller.ex:42]: Suppose I have the seed and address (of pool) both as environment variables. Is it possible to derive the curve from these two informations?

Comment (roychowdhuryrohit-dev) [reply lib/archethic_web/controllers/faucet_controller.ex:42]: Is it possible to determine the curve just from seed/address? Because the private key is not available in environment.

If I can get the curve I can use this :
```elixir
new(type, data, seed, index, curve \\ Crypto.default_curve())
```

If I can get the private key I can use this instead : 
```elixir
new_with_keys(type, data, previous_private_key, previous_public_key, next_public_key)
```

But without either of them I can't generate a new transaction.

File: lib/archethic_web/router.ex (hunk 5 above)
Comment (roychowdhuryrohit-dev) [reply lib/archethic_web/router.ex:28]: Is it possible to get the public key from the seed/address? Without the public key I can't determine the curve.

If I can get the curve I can use this :
```elixir
new(type, data, seed, index, curve \\ Crypto.default_curve())
```

If I can get the private key I can use this instead : 
```elixir
new_with_keys(type, data, previous_private_key, previous_public_key, next_public_key)
```

But without either of them I can't generate a new transaction.

File: lib/archethic_web/controllers/faucet_controller.ex (hunk 3 above)
Comment (roychowdhuryrohit-dev) [reply lib/archethic_web/controllers/faucet_controller.ex:42]: Is it possible to get the public key from the seed/address? Without the public key I can't determine the curve.

If I can get the curve I can use this :
```elixir
new(type, data, seed, index, curve \\ Crypto.default_curve())
```

If I can get the private key I can use this instead : 
```elixir
new_with_keys(type, data, previous_private_key, previous_public_key, next_public_key)
```

But without either of them I can't generate a new transaction.

Comment (samuel-uniris) [reply lib/archethic_web/controllers/faucet_controller.ex:42]: I don't think it's possible.
We should rethink the way to add the testnet seed into the conf.
We could imagine something like:

config/config.exs
```elixir
 config :archethic, ArchEthicWeb.FaucetController,
      seed: "3A7B579DBFB7CEBE26293850058F180A65D6A3D2F6964543F5EDE07BEB2EFDA4" |> Base.decode16!
```

config/dev.exs

```elixir
config  :archethic, ArchEthic.Bootstrap.NetworkInit, 
   genesis_pools: [
      # Address generated from the faucet seed
      %{ address: "00EC64107CA604A6B954037CFA91ED18315A77A94FBAFD91275CEE07FA45EAF893", amount: 10000000}
   ]
  
```

config/prod.exs

```elixir
config :archethic, ArchEthic.Bootstrap.NetworkInit, 
  genesis_pools: [
    ...
    ...,
   (if System.get_env("ARCHETHIC_TESTNET", "false") == "true", do:       
       "00EC64107CA604A6B954037CFA91ED18315A77A94FBAFD91275CEE07FA45EAF893"
    )
  ]
  |> Enum.filter(& &1)
```

Comment (roychowdhuryrohit-dev) [reply lib/archethic_web/controllers/faucet_controller.ex:42]: Suppose I have both the seed and address now. But I don't think I can do a new transaction without knowing the curve which I cannot derive without knowing the public key.

Comment (samuel-uniris) [reply lib/archethic_web/controllers/faucet_controller.ex:42]: You can because the address uses the default curve, as we are not customizing the faucet address

Comment (roychowdhuryrohit-dev) [reply lib/archethic_web/controllers/faucet_controller.ex:42]: But in docsify it is mentioned we can customise the pool address. Should we remove that from doc?

Comment (samuel-uniris) [reply lib/archethic_web/controllers/faucet_controller.ex:42]: Yes we will if it works more

File: assets/css/app.scss (hunk 1 above)
Comment (roychowdhuryrohit-dev) [reply assets/css/app.scss:8]: I was wondering if the JavaScript DOM manipulation works in the context of Server Side Rendering nature of Phoenix. Do I need to re-render the root layout every time any child elements get removed? If that's the case then why does the initial loading of root layout look like this? 

Screenshot 2021-09-14 at 01 30 18

Screenshot 2021-09-14 at 01 25 58

Comment (roychowdhuryrohit-dev) [reply assets/css/app.scss:8]: Using the default Bootstrap alert css for now.

File: lib/archethic_web/controllers/faucet_controller.ex (hunk 3 above)
Comment (roychowdhuryrohit-dev) [reply lib/archethic_web/controllers/faucet_controller.ex:42]: > You can because the address uses the default curve, as we are not customizing the faucet address

Right now the address is `0073bdaf847037115914ff5ca15e52d162db57b5089d5e4bf2005d825592c9c945` which has been derived using secp256r1. I am changing the existing pool address to the one using ed25519 default curve.

**Comment by roychowdhuryrohit-dev** [comment]
_TODO : Remove https://github.com/archethic-foundation/archethic-node/pull/63 from git history using rebase _

File: lib/archethic_web/router.ex (hunk 5 above)
Comment (roychowdhuryrohit-dev) [reply lib/archethic_web/router.ex:28]: In release `Mix.env()` doesn't work as Mix is not available. I want to inject these two lines at only compile time according to Mix.env() condition (:dev or :prod with testnet).
```elixir
get("/faucet", FaucetController, :index)
post("/faucet", FaucetController, :create_transfer)
```
Do I need to leverage Macros to inject something at compile time according to conditions?

Comment (roychowdhuryrohit-dev) [reply lib/archethic_web/router.ex:28]: Using config and not Macros code injection to resolve this issue.

Comment (roychowdhuryrohit-dev) [reply lib/archethic_web/router.ex:28]: Okay `Mix.env() == :dev` was enough for this since in release we can check `System.get_env("ARCHETHIC_NETWORK_TYPE") == "testnet"`.

**Comment by roychowdhuryrohit-dev** [comment]
Closing this PR since https://github.com/archethic-foundation/archethic-node/tree/36cb7d25033be65b79d37e1e84966592fa1e6dc8 is outdated.
